src/zenml/models/v2/core/pipeline_run.py

```diff
@@ -308,6 +310,52 @@ def produced_artifact_versions(self) -> List["ArtifactVersionResponse"]:
 
         return get_artifacts_versions_of_pipeline_run(self, only_produced=True)
 
+    def refresh_run_status(self) -> "PipelineRunResponse":
+        """Method to refresh the status of a pipeline run.
+
+        Returns:
+            The updated pipeline.
+
+        Raises:
+            ValueError: If the stack of the run response is None.
+        """
+        # Check if the stack still accessible
+        if self.stack is None:
+            raise ValueError(
+                "The pipeline run response does not have a stack. It may "
+                "have been deleted or you might not have access to it."
+            )
+
+        # Create the orchestrator instance
+        from zenml.enums import StackComponentType
+        from zenml.orchestrators.base_orchestrator import BaseOrchestrator
+        from zenml.stack.stack_component import StackComponent
+
+        component_model = self.stack.components[
+            StackComponentType.ORCHESTRATOR
+        ][0]
+
+        orchestrator = cast(
+            BaseOrchestrator,
+            StackComponent.from_model(component_model=component_model),
+        )
+
+        # Fetch the status
+        status = orchestrator.fetch_status(run=self)
+
+        # If it is different from the current status, update it
+        if status != self.status:
```

Comment:
I think we could instead check if the run is in initializing/running, and only update the status in this case?
If the run is already in completed/failed, there is no refreshing to be done. Additionally, it might even run into some weird edge cases like a run that failed on the orchestrator side but actually finished successfully for ZenML (maybe something during the teardown failed)

Reply:
Good idea. I added the condition.